RavenDB-7589

ppekrol opened the pull request.

myarichuk committed:
RavenDB-7589 wip

ppekrol committed:
RavenDB-7589 moved identities outside the DatabaseRecord

ppekrol committed:
RavenDB-7589 changed IdentitiesPrefix from db/identities/ to identities/ to avoid possible clash with DatabasePrefix (db/)

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/RestoreBackupTask.cs:
```diff
@@ -139,7 +140,7 @@ public class RestoreBackupTask
                         options |= InitializeOptions.GenerateNewDatabaseId;
                     database.Initialize(options);
-                    SmugglerRestore(_restoreConfiguration.BackupLocation, database, databaseRecord, onProgress, restoreResult);
+                    SmugglerRestore(_restoreConfiguration.BackupLocation, database, databaseRecord, restoreSettings.Identities, onProgress, restoreResult);
```
Need to save identities here. Will do that tomorrow.

ayende commented on src/Raven.Server/Documents/PeriodicBackup/RestoreBackupTask.cs:
```diff
@@ -271,6 +272,7 @@ private List<string> GetFilesForRestore(string backupLocation)
             string backupDirectory, 
             DocumentDatabase database, 
             DatabaseRecord databaseRecord,
+            Dictionary<string, long> identities,
```
What if we have a _lot_ of them?

ayende commented on src/Raven.Server/Documents/Handlers/Debugging/IdentityDebugHandler.cs:
```diff
@@ -14,15 +13,10 @@ public Task GetIdentities()
-                    context.Write(writer, identitiesAsJson);
-                    writer.Flush();
-                }
+                var json = Database.ServerStore.Cluster.Read(context, Constants.Documents.IdentitiesPrefix + Database.Name.ToLowerInvariant());
```
This assume a single object holding all identities per database, right?
That is going to be a problem if we'll have users with thousands of identities. Better to use a value per identity per db.
So the key is going to be `db/db-name/identities/identity-name`.
The idea is that deleting the db will also delete the identities.

ppekrol committed:
RavenDB-7589
- each identity is a separate entry in 'Identities' tree with key 'db/<dbname>/identities/<prefix>'
- added paging to 'debug/identities' endpoint
- on smuggler import or database restore, identities are processed in batches (1024)

ppekrol committed:
RavenDB-7589 do not materialize all identities during smuggling, it can be a huge list

ppekrol commented:
Changes applied.
